File: backend/internal/api/spotify.go
package api

import (
	"context"
	"homehub/internal/spotify"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

// Spotify integration: search/browse via the Web API (the caller's own
// account, PKCE — no client secret), playback via the speakers' linked
// account (see internal/sonos/services.go). Search/browse is open to admins
// and kid profiles: a kid searches as their own connected account, everyone
// else as the household's (see internal/spotify). Connecting an account is
// the same split — a kid's login links their profile's account, an admin's
// the household one — while the developer app's client ID stays admin-only.
// All handlers are nil-safe when the Spotify client isn't wired.

const spotifyTimeout = 10 * time.Second

// spotifyAccount resolves which connected account serves this caller: a kid
// profile always gets its own, everyone else gets the household's.
func (s *Server) spotifyAccount(r *http.Request) *spotify.Account {
	if u := currentUser(r); u != nil && u.Kid {
		return s.Spotify.For(u.ID)
	}
	return s.Spotify.For("")
}

// spotifyRedirectURI computes the OAuth redirect URI to use for the origin
// the request arrived on. It must be registered verbatim in the Spotify
// app, so the status endpoint surfaces it for the user to copy.
//
// Spotify only accepts HTTPS or loopback redirect URIs. Over HTTPS the
// URI points back at this server and the flow completes automatically.
// Over plain HTTP (the common LAN setup) a parked loopback URI is used
// instead: the browser can't load it, but the authorization code is in the
// address bar and the user pastes that address back into the Music view
// (see spotifyExchange). manual reports which of the two flows applies.
func spotifyRedirectURI(r *http.Request) (uri string, manual bool) {
	if isSecureRequest(r) {
		host := r.Host
		if xfh := r.Header.Get("X-Forwarded-Host"); xfh != "" {
			host = xfh
		}
		return "https://" + host + "/api/spotify/callback", false
	}
	// Match the real listen port so that a browser running on the HomeHub
	// host itself still completes automatically.
	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}
	return "http://127.0.0.1:" + port + "/api/spotify/callback", true
}

func (s *Server) requireSpotify(w http.ResponseWriter) bool {
	if s.Spotify == nil {
		writeError(w, http.StatusServiceUnavailable, "Spotify integration is not available")
		return false
	}
	return true
}

// spotifyStatus handles GET /api/spotify/status — the caller's own
// account's state (a kid's login is a different connection from the
// household's).
func (s *Server) spotifyStatus(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	st := s.spotifyAccount(r).Status()
	uri, manual := spotifyRedirectURI(r)
	writeJSON(w, http.StatusOK, map[string]interface{}{
		"configured":   st.Configured,
		"connected":    st.Connected,
		"display_name": st.DisplayName,
		"redirect_uri": uri,
		"manual":       manual,
		// Whether this login can start Connect playback (KEF speakers). A
		// login made before HomeHub asked for the player scopes searches
		// fine and can't play, and only a reconnect fixes that — so the
		// frontend has to be able to tell the difference.
		"playback": st.Playback,
	})
}

// spotifySetConfig handles PUT /api/spotify/config with {"client_id": "..."}.
func (s *Server) spotifySetConfig(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	var body struct {
		ClientID string `json:"client_id"`
	}
	if !decodeBody(w, r, &body) {
		return
	}
	if strings.TrimSpace(body.ClientID) == "" {
		writeError(w, http.StatusBadRequest, "client_id is required")
		return
	}
	if err := s.Spotify.SetClientID(body.ClientID); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// spotifyLogin handles GET /api/spotify/login — returns the authorize URL
// the frontend should navigate to. Which account the flow connects is the
// caller's: a kid's login links the kid profile's account, an admin's the
// household one.
func (s *Server) spotifyLogin(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	uri, _ := spotifyRedirectURI(r)
	key := ""
	if u := currentUser(r); u != nil && u.Kid {
		key = u.ID
	}
	u, err := s.Spotify.AuthURL(key, uri)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"url": u})
}

// spotifyCallback handles GET /api/spotify/callback — the browser lands
// here from Spotify's consent page. On success it bounces back into the
// app — the Music view for the household account, the kid app for a kid's
// own; errors are shown by redirecting with a query the view toasts.
func (s *Server) spotifyCallback(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	q := r.URL.Query()
	if e := q.Get("error"); e != "" {
		http.Redirect(w, r, "/#/music?spotify_error="+e, http.StatusFound)
		return
	}
	code, state := q.Get("code"), q.Get("state")
	if code == "" || state == "" {
		writeError(w, http.StatusBadRequest, "missing code/state")
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), spotifyTimeout)
	defer cancel()
	accountKey, err := s.Spotify.HandleCallback(ctx, code, state)
	if err != nil {
		http.Redirect(w, r, "/#/music?spotify_error="+err.Error(), http.StatusFound)
		return
	}
	if accountKey != "" {
		// A kid profile connected its own account — land back on the kid
		// app, which has no Music route to toast into.
		http.Redirect(w, r, "/", http.StatusFound)
		return
	}
	http.Redirect(w, r, "/#/music?spotify=connected", http.StatusFound)
}

// spotifyExchange handles POST /api/spotify/exchange — the paste-the-URL
// finish for the manual (plain-HTTP) flow. Body: {"url": "<address the
// browser landed on after consent>"}.
func (s *Server) spotifyExchange(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	var body struct {
		URL string `json:"url"`
	}
	if !decodeBody(w, r, &body) {
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), spotifyTimeout)
	defer cancel()
	if _, err := s.Spotify.ExchangeRedirect(ctx, body.URL); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// spotifyDisconnect handles POST /api/spotify/disconnect — drops the
// caller's own account's tokens; every other account is untouched.
func (s *Server) spotifyDisconnect(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	if err := s.spotifyAccount(r).Disconnect(); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// spotifySearch handles GET /api/spotify/search?q=…&limit=…&kind=…&offset=…
//
// `kind` narrows the search to one of tracks/albums/playlists/artists and
// `offset` pages into it — together they are what a shelf's "Show more"
// needs, since Spotify caps a search's limit at 10 and paging is the only
// way to an eleventh result. Both are optional: without them this is the
// broad four-kind search every first query makes.
func (s *Server) spotifySearch(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	q := strings.TrimSpace(r.URL.Query().Get("q"))
	if q == "" {
		writeError(w, http.StatusBadRequest, "q is required")
		return
	}
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	offset, _ := strconv.Atoi(r.URL.Query().Get("offset"))
	kind := r.URL.Query().Get("kind")
	ctx, cancel := context.WithTimeout(r.Context(), spotifyTimeout)
	defer cancel()
	res, err := s.spotifyAccount(r).SearchPage(ctx, q, kind, limit, offset)
	if err != nil {
		writeError(w, spotifyErrStatus(err), err.Error())
		return
	}
	writeJSON(w, http.StatusOK, res)
}

// spotifyPlaylists handles GET /api/spotify/playlists — the connected
// account's own playlists, for browsing without typing.
func (s *Server) spotifyPlaylists(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), spotifyTimeout)
	defer cancel()
	items, err := s.spotifyAccount(r).MyPlaylists(ctx, 30)
	if err != nil {
		writeError(w, spotifyErrStatus(err), err.Error())
		return
	}
	writeJSON(w, http.StatusOK, items)
}

// spotifyArtist handles GET /api/spotify/artist?uri=spotify:artist:… — an
// artist's page: top tracks and albums, for the screen behind a search
// result (DESIGN.md §15).
func (s *Server) spotifyArtist(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	uri := strings.TrimSpace(r.URL.Query().Get("uri"))
	if uri == "" {
		writeError(w, http.StatusBadRequest, "uri is required")
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), spotifyTimeout)
	defer cancel()
	det, err := s.spotifyAccount(r).Artist(ctx, uri)
	if err != nil {
		writeError(w, spotifyErrStatus(err), err.Error())
		return
	}
	writeJSON(w, http.StatusOK, det)
}

// spotifyContext handles GET /api/spotify/context?uri=spotify:playlist:…
// or spotify:album:… — the tracks inside a playlist or album, for the
// screen behind a favorite that turns out to be a list rather than one song.
func (s *Server) spotifyContext(w http.ResponseWriter, r *http.Request) {
	if !s.requireSpotify(w) {
		return
	}
	uri := strings.TrimSpace(r.URL.Query().Get("uri"))
	if uri == "" {
		writeError(w, http.StatusBadRequest, "uri is required")
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), spotifyTimeout)
	defer cancel()
	det, err := s.spotifyAccount(r).Context(ctx, uri)
	if err != nil {
		writeError(w, spotifyErrStatus(err), err.Error())
		return
	}
	writeJSON(w, http.StatusOK, det)
}

// spotifyErrStatus maps "not connected" to 409 so the frontend can prompt
// re-auth, everything else to bad-gateway.
func spotifyErrStatus(err error) int {
	if strings.Contains(err.Error(), "not connected") {
		return http.StatusConflict
	}
	return http.StatusBadGateway
}

File: frontend/src/lib/api/spotify.ts
/**
 * The catalog, searched and browsed through the account's own
 * PKCE grant.
 *
 * One slice of the flat `api` object — lib/api.ts spreads these together,
 * so every call site's `api.foo()` is unchanged. Split by domain because a
 * thousand-line object is not a thing anyone reads.
 */

import { req, json } from "./http";
import type {
  SpotifyArtistDetail,
  SpotifyContextDetail,
  SpotifyItem,
  SpotifyLibrary,
  SpotifyListening,
  SpotifyResults,
  SpotifyStatus,
} from "../types";

export const spotifyApi = {
  // Spotify search/browse (user's own account via PKCE — configured in the Music view)
  spotifyStatus() { return req<SpotifyStatus>("/spotify/status"); },
  spotifySetConfig(clientId: string) {
    return req<void>("/spotify/config", { method: "PUT", body: json({ client_id: clientId }) });
  },
  spotifyLoginURL() { return req<{ url: string }>("/spotify/login"); },
  // Manual-flow finish: pass the full address the browser landed on after consent.
  spotifyExchange(url: string) {
    return req<void>("/spotify/exchange", { method: "POST", body: json({ url }) });
  },
  spotifyDisconnect() { return req<void>("/spotify/disconnect", { method: "POST" }); },
  // `kind` narrows to one of tracks/albums/playlists/artists and `offset`
  // pages into it — what a shelf's "Show more" needs, since Spotify caps a
  // search at ten results per kind. `signal` lets a superseded search be
  // called off rather than left to finish and be discarded.
  spotifySearch(
    q: string,
    limit = 8,
    opts: { kind?: string; offset?: number; signal?: AbortSignal } = {},
  ) {
    const p = new URLSearchParams({ q, limit: String(limit) });
    if (opts.kind) p.set("kind", opts.kind);
    if (opts.offset) p.set("offset", String(opts.offset));
    return req<SpotifyResults>(`/spotify/search?${p}`, { signal: opts.signal });
  },
  spotifyMyPlaylists() { return req<SpotifyItem[]>("/spotify/playlists"); },
  // What the account has been playing, for the idle shelves. 409 means the
  // login predates the listening scopes — a reconnect, not a fault.
  spotifyListening() { return req<SpotifyListening>("/spotify/listening"); },
  // Songs to continue with, seeded from an artist name — the same engine
  // "play similar" uses when a queue runs dry, asked for on purpose. By
  // name because that is what a speaker reports about what it is playing.
  spotifySimilar(artist: string, limit = 8) {
    return req<SpotifyItem[]>(
      `/spotify/similar?artist=${encodeURIComponent(artist)}&limit=${limit}`,
    );
  },
  // An artist's page — top tracks and albums, behind a search result.
  spotifyArtist(uri: string) {
    return req<SpotifyArtistDetail>(`/spotify/artist?uri=${encodeURIComponent(uri)}`);
  },
  // A playlist or album's own tracks — behind a favorite that turns out to
  // be a list rather than one song.
  // Whether a track is in the account's library, and putting it there. The
  // read needs no scope the login didn't already have; the write does, which
  // is why status.library exists — hide the control rather than offer a tap
  // that will be refused.
  spotifySaved(uri: string) {
    return req<{ saved: boolean }>(`/spotify/saved?uri=${encodeURIComponent(uri)}`);
  },
  spotifySetSaved(uri: string, saved: boolean) {
    return req<void>("/spotify/saved", { method: "PUT", body: json({ uri, saved }) });
  },
  spotifyContext(uri: string) {
    return req<SpotifyContextDetail>(`/spotify/context?uri=${encodeURIComponent(uri)}`);
  },
  // The rest of the collection — saved albums, kept playlists, the artists
  // this account actually plays — read as three shelves in one round trip.
  // A shelf whose scope was refused comes back empty rather than failing the
  // request; only all three failing is the account's problem rather than one
  // permission's.
  spotifyLibrary(limit = 20) {
    return req<SpotifyLibrary>(`/spotify/library?limit=${limit}`);
  },
  // What came out lately. The one shelf that is about the catalog rather
  // than about this household, and so the only answer available on an
  // evening when nobody wants to hear anything they already know.
  spotifyNewReleases(limit = 20) {
    return req<SpotifyItem[]>(`/spotify/new-releases?limit=${limit}`);
  },
};
